# -*- snakemake -*-
include: '../settings.rules'

DEFAULT_RULES = ['rsem_calculate_expression']

config_default = { 
    'bio.ngs.rnaseq.rsem' : {
        'rules' : DEFAULT_RULES,
        'annot_gtf' : config['bio.ngs.settings']['annotation']['transcript_annot_gtf'],
        'threads' : config['bio.ngs.settings']['threads'],
        'ref' : config['bio.ngs.settings']['db']['ref'],
        'ref_sfx' : '.transcripts.fa',
        'index' : "",
        'calculate-expression' : {
            'cmd' : 'rsem-calculate-expression',
            'options' : '--no-bam-output',
            'bowtie-options' : "--no-bam-output --bowtie-chunkmbs 512",
        },
    },
}

config = update_config(config_default, config)

try:
    from snakemakelib.bio.ngs.db import index, annotation
    # Update annotation
    config['bio.ngs.rnaseq.rsem']['annot_gtf'] = annotation(
        annotation=config['bio.ngs.rnaseq.rsem']['annot_gtf'],
        db_cfg=config['bio.ngs.settings']['db'])
    # Update index
    config['bio.ngs.rnaseq.rsem']['index'] = index(
        ref = config['bio.ngs.rnaseq.rsem']['ref'],
        index = config['bio.ngs.rnaseq.rsem']['index'],
        application = 'rsem',
        build = config['bio.ngs.settings']['db']['build'])
except:
    pass

for rule in config['bio.ngs.rnaseq.rsem']['rules']:
    include: os.path.join("rsem", rule + ".rule")
