# -*- snakemake -*-
from snakemakelib.bio.ngs.methylseq.bismark import report_label, align_suffix

include: '../settings.rules'

DEFAULT_RULES = [
    'bismark_report',
    'bismark_PE_align',
    'bismark_deduplicate_bismark',
    'bismark_genome_preparation',
    'bismark_merge_alignment_reports',
    'bismark_methylation_extractor'
]

config_default = {
    'bio.ngs.methylseq.bismark' : {
        'rules' : DEFAULT_RULES,
        'bowtie2' : True,
        'ref' : "",
        'trim_galore' : True,
        'paired_end' : True,
        'align' : {
            'cmd' : 'bismark',
            'options' : '--bam --phred33-quals',
            'threads' : 8,
        },
    },
}

config = update_config(config_default, config)

if config['bio.ngs.methylseq.bismark']['trim_galore']:
    include: '../qc/trim_galore.rules'

for rule in config['bio.ngs.methylseq.bismark']['rules']:
    include: os.path.join("bismark", rule + ".rule")
