# -*- snakemake -*-
#
# GEM: ChIP-seq and ChIP-exo analysis tool
# Genome wide Event finding and Motif discovery 
#
# http://www.psrg.csail.mit.edu/gem/
#
include: '../settings.rules'

DEFAULT_RULES = ['gem_run']

GEM_JAR_PROGRAM = "gem.jar"

config_default = {
    'bio.ngs.motif.gem' : {
        'rules' : DEFAULT_RULES,
        'home' : os.getenv("GEM_MOTIF_HOME", os.curdir),
        'java_mem' : config['bio.ngs.settings']['java']['java_mem'],
        'java_tmpdir' : config['bio.ngs.settings']['java']['java_tmpdir'],
        'options' : '',
    },
}

update_config(config_default, config)
config = config_default

config_default2 = {
    'bio.ngs.motif.gem' : {
        'read_distribution': os.path.join(config['bio.ngs.motif.gem']['home'], 'Read_Distribution_default.txt'),
        'jar' : os.path.join(config['bio.ngs.motif.gem']['home'], GEM_JAR_PROGRAM),
    },
}

update_config(config_default2, config)
config = config_default2

config_default3 = {
    'bio.ngs.motif.gem' : {
        'cmd' : "java -Xmx" + config['bio.ngs.motif.gem']['java_mem'] + " -Djava.io.tmpdir=" + config['bio.ngs.motif.gem']['java_tmpdir'] +  " -jar " + config['bio.ngs.motif.gem']['jar'],
    },
}

update_config(config_default3, config)
config = config_default3

for rule in config['bio.ngs.motif.gem']['rules']:
    include: os.path.join("gem", rule + ".rule")
