# -*- snakemake -*-
include: '../settings.rules'
# Include required picard rules; do we only need the index rule?
include: '../qc/picard.rules'

GATK_JAR_PROGRAM = "GenomeAnalysisTK.jar"

DEFAULT_RULES = ['gatk_base_recalibrator',
                 'gatk_clip_reads',
                 'gatk_combine_variants',
                 'gatk_indel_realigner',
                 'gatk_print_reads',
                 'gatk_read_backed_phasing',
                 'gatk_realigner_target_creator',
                 'gatk_select_indel_variants',
                 'gatk_select_snp_variants',
                 'gatk_select_variants',
                 'gatk_select_variants_region',
                 'gatk_select_variants_sample',
                 'gatk_unified_genotype',
                 'gatk_variant_JEXL_filtration',
                 'gatk_variant_eval',
                 'gatk_variant_indel_JEXL_filtration',
                 'gatk_variant_snp_JEXL_filtration']


config_default = { 
    'bio.ngs.tools.gatk' : {
        'rules' : DEFAULT_RULES,
        'bait_regions' : config['bio.ngs.settings']['sequence_capture']['bait_regions'],
        'bam_list' : "",
        'cov_interval' : "regional",
        'dbsnp' : config['bio.ngs.settings']['db']['dbsnp'],
        'home' : os.getenv("GATK_HOME", os.curdir),
        'java_mem' : config['bio.ngs.settings']['java']['java_mem'],
        'java_tmpdir' : config['bio.ngs.settings']['java']['java_tmpdir'],
        'ref' : config['bio.ngs.settings']['db']['ref'],
        'target_regions' : config['bio.ngs.settings']['sequence_capture']['target_regions'],
        'threads' : config['bio.ngs.settings']['threads'],
        'vcfsuffix' : ".vcf",
    },
}

update_config(config_default, config)
config = config_default

config_default2 = {'bio.ngs.tools.gatk': {}}

config_default2 = {
    'bio.ngs.tools.gatk' : {
        'jar' : os.path.join(config['bio.ngs.tools.gatk']['home'], GATK_JAR_PROGRAM),
        'known_sites' : config['bio.ngs.tools.gatk']['dbsnp'],
    },
}

update_config(config_default2, config)
config = config_default2

config_default3 = {
    'bio.ngs.tools.gatk': {
        'cmd' : "java -Xmx" + config['bio.ngs.tools.gatk']['java_mem'] + " -Djava.io.tmpdir=" + config['bio.ngs.tools.gatk']['java_tmpdir'] +  " -jar " + config['bio.ngs.tools.gatk']['jar'],
    }
}

update_config(config_default3, config)
config = config_default3

for rule in config['bio.ngs.tools.gatk']['rules']:
    include: os.path.join("gatk", rule + ".rule")
