<!-- -*- jinja2 -*- -->
<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <title>Single-cell RNA-Seq summary report</title>

    {{ bokeh_css }}
    {{ bokeh_js }}
    
    <!-- Include bokeh script commands -->
    {% if rseqc is defined %}
      {{ rseqc['fig']['script'] | indent(8) }}
    {% endif %}

    {% if star is defined %}
      {{ star['fig']['script'] | indent(8) }}
      {% if star['table'] is not none %}
	{{ star['table']['script'] | indent(8) }}
      {% endif %}
    {% endif %}

    {% if align is defined %}
      {{ align['fig']['script'] | indent(8) }}
      {{ align['table']['script'] | indent(8) }}
    {% endif %}

    {% if rpkmforgenes is defined %}
      {{ rpkmforgenes['fig']['script'] | indent(8) }}
    {% endif %}

    {% if rsem is defined %}
      {{ rsem['fig']['script'] | indent(8) }}
    {% endif %}
    
  </head>
  <body>

    <h1>scRNA-seq workflow</h1>
    
    The current single cell RNA-seq workflow does the following steps:
    
    <ol>
      <li>Alignment with star</li>
      <li>Quality control with RSeQC</li>
      <li>Quantification with RSEM and/or rpkmforgenes.py</li>
    </ol>

    {% include 'snakemake.html' ignore missing %}

    {% include 'workflow.html' ignore missing %}

    <h1>Single-cell RNA-seq QC summary</h1>

    <h2>Run-level QC metrics</h2>

    <h2>Sample-level QC metrics</h2>

    {% include 'workflow_scrnaseq_qc_align.html' ignore missing %}

    {% include 'workflow_scrnaseq_qc_quantification.html' ignore missing %}
    
  </body>
</html>
